### ISSUE TITLE
copyContainer does not copy all files of a directory

### ISSUE THREAD

**Issue by jjohnstn** [spotify/docker-client#547]
## Description

I am using docker-client 3.4.0 and am trying to use it to copy files from a container to the host.  If I choose a directory with a number of files (including other directories), I only get some of the files in the archive stream that is returned.  I am using code similar to that found at: 

http://www.programcreek.com/java-api-examples/index.php?api=org.apache.commons.compress.archivers.tar.TarArchiveEntry

When I do this, I see the main directory and then I read a number of files, then there are no more tar entries, but the directory has not completely been copied.  For a small directory with, for example, 4 files, it works fine.

Is this a known issue or is there an upper limit for the size of data copied or is this something that is being caused by the daemon?  I am using RHEL 7.2 and using the docker-common-1.10.3-46.el7.14.x86_64 rpm.  For apche commons compress, I am using 1.6.0.

## How to reproduce

Start up the docker daemon on RHEL 7.2 and then run for example:

copyContainer(CONTAINER_ID, "/bin")

where the directory to copy has a large number of files and/or other directories.

Untar the results as mentioned above using org.apache.commons.compress.TarArchiveInputStream

## What do you expect

I expect "/bin" and all files from /bin to show up represented by TarArchiveEntry instances.

## What happened instead

I only see a truncated number of files show up in the TarArchiveInputStream.  In one case, it stopped after .gitignore.

## Software:

- `docker version`: [Docker version 1.10.3, build 5206701-unsupported]
- Spotify's docker-client version: [3.4.0]

## Full backtrace

N/A
```

**Comment by jjohnstn** [comment]
I have solved the problem.  It appears that the InputStream from copyContainer sometimes returns less bytes than asked for (likely has an available() method that does not default to return 0).  When processing a TarArchiveInputStream that uses this stream as it's source, sometimes the read returns less than the asked for bytes while processing a tar entry header.  This causes the TarArchiveInputStream to return null for getNextEntry() and then we attempt to close the stream.  Unfortunately, this doesn't close because in the low levels there is an uncloseable input stream.  I think that some form of locks are still held and any subsequent requests go into park and check and just sit there indefinitely.  This freezes any future docker daemon requests including those from the command line.  The solution is to wrap the InputStream with a basic class that implements InputStream (read() can just call the copyContainer's InputStream read()).  This sets the default available() method to always return 0 and thus the reads get blocked until either they are filled or the source returns EOF.  When the whole file is processed correctly, there are no issues, all files get copied, and nothing hangs.  Requesting copying of a small file or small amount of files doesn't reproduce as I assume that the full header reads are always satisfied.

**Comment by mattnworb** [comment]
Thanks for the report. Do you have a test case that reproduces this? For instance if you could create a repo or gist with a pom.xml with all of the library versions you are using, a directory with a files where this copy happens for, etc., then that would be really useful.

**Comment by rgrunber** [comment]
This isn't a docker-client bug, but rather a bug on our end for using commons-compress 1.6. Feel free to close.

I was trying everything to reproduce this since but no matter what I did, the copyContainer/copyArchive worked perfectly in the spotify/docker-client tests.

Then after staring at commons-compress 1.6 code in TarArchiveInputStream : 

    byte[] record = new byte[recordSize];

    int readNow = is.read(record);
    count(readNow);
    if (readNow != recordSize) {
      return null;
    }

There's no guarantee that recordSize bytes will always be read. So if commons-compress 1.6 has issues the only thing to test is whether an updated version fixed the problem (since docker-client uses commons-compress 1.9 by default) Looking through past commits : https://github.com/apache/commons-compress/commit/2221ee87b406cf1781f0a4f18914812db2256b01#diff-f294260f85c2ae2a0c5b3a005d5f9f04 , which was fixed 1.7.
